### ISSUE TITLE
Add LSTM-based traffic prediction

### ISSUE THREAD

**Issue by qarawlus** [RealVNF/coord-sim#163]
I have taken a look at LSTM prediction. Here is how I think we can implement it:

Since we need training data, I suggest we do the following:
- Get the inter arrival mean from the simulator for the run. This would mean that we would be able to run with all scenarios  including traces and MMPP. 
- Use the first run with a new inter arrival mean to calculate an oracle prediction
- Pass this list of oracle predictions to the LSTM model as a training data
- Once the LSTM is trained, we can use it to predict our next arrival time using the data from the inter arrival mean
- To make sure we predict arrival times that are less than run duration, we can use the same loop conditions as we have in our current traffic prediction model. This will make sure that we do not generate more flow arrivals than we need.

Once we have enough predicted arrival times (complete list), we pass that back to the simulator interface, we calculate the traffic data rate for that run based on the predicted list, and pass that to the agent.

We keep the same trained LSTM model as long as the inter arrival mean is the same. If the inter arrival mean changes, we repeat from the beginning and train again with the new arrival mean.

**Comment by stefanbschneider** [comment]
I'm not sure I can fully follow. The goal of the LSTM prediction is, similar to what the oracle does now, predict the mean inter-arrival time (or is it the expected total flow volume?) of the next run, right?
So as input, we pass the arrival times (or volumes?) of the last X runs and the LSTM predicts it for the next run.

I would decouple training and testing/inference. Training should be much easier and faster if we do it up front directly on the data (= simulator traces) without any simulator logic in between.
Then again, inference should be quite simple: Just keep track of the last X runs (we do that anyway) and pass it as input to the LSTM.
The output, ie, prediction, we pass to the agent; similar to how we now pass the oracle prediction.

Of course, it would be nice, to also support non-trace-driven traffic or online learning, but I think those would make things considerably more difficult without bringing major benefits. (If anything, I'd think about online learning; but I really don't think we need to support non-trace traffic with LSTM)

**Comment by qarawlus** [comment]
What I have in mind will allow us to use LSTMs to predict the traffic arrival list of flow arrival times that we currently use in the simulator. We will not predict the inter arrival mean, since the mean is now only used to calculate the arrival list. Instead, I suggest we predict the exact flow arrival times based on the mean we currently have.

The reason I chose this approach is because it will have minimum changes to the way the simulator works now and would actually be more simple to implement. 

Granted, there will be some performance degradation because we now add LSTM training online, but since our prediction model will be simple and we only retrain after mean changes, I expect these to performance effects to be small. I can also mitigate these effects by using MPI to have another processor do the prediction.

**Comment by stefanbschneider** [comment]
Why do you want to predict exact flow arrival times? What would you do with it? The LSTM shouldn't change the way the simulator works. It should only provide predicted future traffic for the RL agent. And the RL agent only observes aggregated traffic per run, not individual flow arrivals.

The goal should be to learn the traffic pattern over multiple runs, eg, day/night cycles in the traffic traces, so that the LSTM can predict if/how traffic is going to increase/decrease in the next run. This only makes sense for traces where the mean changes over time. With fixed mean, we will still have some deviations from the mean due to Poisson arrival, but that's nothing that could be predicted (other than guessing the mean) - arrival of future flows is completely independent from previous flows.

**Comment by qarawlus** [comment]
> RL agent only observes aggregated traffic per run, not individual flow arrivals

Exactly. I want to pass the predicted arrival list to the simulator interface, where it will calculate the predicted aggregated traffic, and pass that to the RL agent instead of the oracle prediction we currently have. 

> The goal should be to learn the traffic pattern over multiple runs, eg, day/night cycles in the traffic traces, so that the LSTM can predict if/how traffic is going to increase/decrease in the next run

This can be easily done as well. We can pass a single mean to generate training data based on that, or we can pass multiple mean (e.g. from a trace file) and generate our training data based on that. 

Some pseudo code of how this might work
```python

# inter arr means dict - {time: arrival_mean}
inter_arrival_means = {0: 10, 1200: 6, 3000: 10}

# training data list
arrival_times = []
run_duration = 100

for time, mean in inter_arrival_means:
    #  same loop as arrival list generation in simulator
    
    while next_arrival < next_run:
        new_flow_arrival_time = random(mean)
        arrival_times.append(new_flow_arrival_time)

# now we have training data for all means

# build and train LSTM model

model = LSTM(nn_config)
model.fit(training_data)

predicted_run_traffic = []

# same loop condition from simulator arrival list generation
while next_arrival < next_run:
    predicted_run_traffic.append(model.predict())


# pass predicted run traffic array back to simulator
# simulator interface calculates total aggregated traffic from the predicted list, passes to RL. 

```

**Comment by qarawlus** [comment]
Status update: 
Implementation done in #165 and testing is underway. 

To-Do: Check accuracy of the last prediction at the end of each run. 

First results available from a test/first scenario with real world traces: https://github.com/RealVNF/evaluation/blob/master/future-traffic/lstm/eval.ipynb

**Comment by qarawlus** [comment]
I will close this for now and move further discussions to a separate issue in the evaluation repo, and will reopen in case of bugs.

**Comment by stefanbschneider** [comment]
Having decided to just predict total aggregated traffic per run, the generate data function seems overly complicated: https://github.com/RealVNF/coord-sim/blob/e3b8af485ee20a2c2cb2f352e5209a53782cff1f/src/coordsim/traffic_predictor/lstm_predictor.py#L144

Why not just something like this?

```python
num_flows = run_duration / inter_arrival_time
agg_traffic = num_flows * flow_size
```

You're generating flows according to the config, but the average/expected traffic can be calculated directly from the config/trace.
Of course, the real traffic may deviate from the mean if we have Poisson traffic.
I'm not sure if training on such randomly generated and noisy traffic helps the LSTM learn to deal with noisy traffic in practice. In practice, the random deviation will most likely be different than in the generated training data - so it may just be missleading.

**Comment by stefanbschneider** [comment]
If no LSTM is configured, that's the oracle prediction, right? https://github.com/RealVNF/coord-sim/blob/e3b8af485ee20a2c2cb2f352e5209a53782cff1f/src/coordsim/traffic_predictor/traffic_predictor.py#L41

Would be nice to wrap that into a similar class like the LSTM predictor, `OraclePredictor`, with the same interface that's called from there.

**Comment by qarawlus** [comment]
> Having decided to just predict total aggregated traffic per run, the generate data function seems overly complicated:
> 
> https://github.com/RealVNF/coord-sim/blob/e3b8af485ee20a2c2cb2f352e5209a53782cff1f/src/coordsim/traffic_predictor/lstm_predictor.py#L144
> 
> Why not just something like this?
> 
> ```python
> num_flows = run_duration / inter_arrival_time
> agg_traffic = num_flows * flow_size
> ```
> 
> You're generating flows according to the config, but the average/expected traffic can be calculated directly from the config/trace.
> Of course, the real traffic may deviate from the mean if we have Poisson traffic.
> I'm not sure if training on such randomly generated and noisy traffic helps the LSTM learn to deal with noisy traffic in practice. In practice, the random deviation will most likely be different than in the generated training data - so it may just be missleading.

We have the option for both. The function has a check that checks if poisson prediction is enabled or not. If it is enabled, then it generates real arrival data based on a Poisson process. But if it is disabled, then it generates the training data exactly how you described. 

https://github.com/RealVNF/coord-sim/blob/e3b8af485ee20a2c2cb2f352e5209a53782cff1f/src/coordsim/traffic_predictor/lstm_predictor.py#L188

**Comment by qarawlus** [comment]
I have made several modification to the LSTM module:
- Training is now done on separate training and testing data. 
- Prediction are now based on simulator data
- The module is trained on a data set that has 0 as the first requested data rate to let the NN know the init data rate of the first run. 

I have also made changes to the way the NN trains, namely the removal of the for loop in the LSTM example and using the built-in epoch counter of keras. I have adjusted the neuron size to be 10 instead of 4 and training for 3000 epochs instead of 10. 

Before the adjusted training method, the NN was not learning much: here is an example of training for 3000 epochs
![image](https://user-images.githubusercontent.com/44025202/91715544-23661f00-eb8e-11ea-930b-06b9f744ceba.png)

After adjusting the training method and using keras's built-in epoch setting without resetting states after each epoch:
![image](https://user-images.githubusercontent.com/44025202/91715586-37aa1c00-eb8e-11ea-84a0-f735b9a91de9.png)


Note that both these NNs were trained on 90% of the available traffic data. 

There is still the issue that the NN does not always predict spikes one step ahead, rather as it happens. I have been continuously looking for a way to address this, but all my tuning attempts have been so far unsuccessful.
